File: jev-experiments/capability-atlas-2026-09-22/runtime-probes.ts
/** Provider-free audit probes. All HTTP responses below are authored fixtures. */
import { validate } from "../experience-prototypes/server/gateway";
import { validateRequest } from "../roadmap/runtime/contract";
import { createJevAdapter } from "../roadmap/runtime/jev";
import { compile } from "../adapters/typescript/index";

const variants = {
  plainNoul: { type: "noul", instructions: "Is it eligible?" },
  structuredInstruction: { type: "choice", instructions: { question: "Which?", facts: [1, 2] }, criteria: { a: "A", b: "B" } },
  structuredChoice: { type: "choice", instructions: "Which?", criteria: { a: { meaning: "A" }, b: { meaning: "B" } } },
  structuredScore: { type: "score", instructions: "How much?", criteria: [{ meaning: "low" }, { meaning: "high" }] },
  noulCriteria: { type: "noul", instructions: "Is it eligible?", criteria: { true: "All rules hold", false: "A rule fails" } },
  nullInstruction: { type: "choice", instructions: null, criteria: { a: "A", b: "B" } },
  nullChoiceDescription: { type: "choice", instructions: "Which?", criteria: { a: null, b: "B" } },
  scoreEleven: { type: "score", instructions: "Which level?", criteria: Array.from({ length: 11 }, (_, i) => String(i)) },
};
const gatewayValidation: Record<string, string> = {};
for (const [id, q] of Object.entries(variants)) {
  try {
    validate({ state: "Example", questions: { q } });
    gatewayValidation[id] = "accepted";
  } catch (error) {
    gatewayValidation[id] = error instanceof Error ? error.message : "rejected";
  }
}
// Untrusted JSON can contain fields excluded by the TypeScript static type.
const request: any = {
  schemaVersion: "1", requestId: "probe", state: { text: "Example" },
  questions: [
    { id: "q", kind: "ordinal", prompt: "Which level?", min: 10, max: 30, step: 10, criteria: [{ meaning: "low" }, { meaning: "medium" }, { meaning: "high" }] },
    { id: "b", kind: "boolean", prompt: "Is it eligible?", criteria: { true: { rules: ["a", "b"] }, false: "Any failure" } },
    { id: "c", kind: "choice", prompt: "Which option?", options: [{ id: "a", label: "A", description: "Detail A" }, { id: "b", label: "B" }] },
  ],
};
let wire: any;
const fixtureFetch = (async (_url: unknown, init: any) => {
  wire = JSON.parse(init.body);
  return new Response(JSON.stringify({ model: "typesafe-ai/jev", answers: {
    q: { type: "score", score: 1.6, confidence: 0.5, legend: { "0": "low", "1": "medium", "2": "high" }, probabilities: { "0": 0.1, "1": 0.2, "2": 0.7 } },
    b: { type: "noul", noul: 0.6 },
    c: { type: "choice", choice: "a", confidence: 0.4, probabilities: { a: 0.7, b: 0.3 } },
  } }), { status: 200, headers: { "Content-Type": "application/json" } });
}) as typeof fetch;
const output = await createJevAdapter("mock-only", fixtureFetch).decide(request);
const schema = {
  type: "object", required: ["rating"], properties: {
    rating: { type: "number", minimum: 0, maximum: 1, description: "Severity?", "x-jev-levels": ["routine", "severe"] },
  },
};
console.log(JSON.stringify({
  networkCalls: 0,
  gatewayValidation,
  sharedExtraFieldsIssues: validateRequest(request),
  sharedElevenLevelIssues: validateRequest({ ...request, questions: [{ id: "eleven", kind: "ordinal", prompt: "Which level?", min: 0, max: 10 }] }),
  mockedWireQuestions: wire.questions,
  adaptedDecisions: output.decisions,
  twoLevelScoreCompilation: compile(schema),
}, null, 2));

File: jev-experiments/capability-atlas-2026-09-22/runtime-probes.py
"""Provider-free validation probes; does not construct a model or API client."""
import importlib.util
import json
from pathlib import Path
import sys

LAB = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(LAB / "src"))
from jev_lab.core import Question

cases = {
    "plainNoul": {"type": "noul", "instructions": "Eligible?"},
    "structuredInstruction": {"type": "noul", "instructions": {"question": "Eligible?"}},
    "structuredChoice": {"type": "choice", "instructions": "Which?", "criteria": {"a": {"meaning": "A"}, "b": "B"}},
    "structuredScore": {"type": "score", "instructions": "How much?", "criteria": [{"meaning": "low"}, {"meaning": "high"}]},
    "noulCriteria": {"type": "noul", "instructions": "Eligible?", "criteria": {"true": "All rules hold", "false": "Any fails"}},
    "nullInstruction": {"type": "choice", "instructions": None, "criteria": {"a": "A", "b": "B"}},
    "nullChoiceDescription": {"type": "choice", "instructions": "Which?", "criteria": {"a": None, "b": "B"}},
    "scoreEleven": {"type": "score", "instructions": "Which?", "criteria": list(map(str, range(11)))},
}
checks = {}
for name, question in cases.items():
    try:
        Question.model_validate(question)
        checks[name] = "accepted"
    except Exception as error:
        checks[name] = "rejected: " + type(error).__name__
spec = importlib.util.spec_from_file_location("jev_local", LAB / "roadmap/mac/jev_local.py")
local = importlib.util.module_from_spec(spec)
spec.loader.exec_module(local)
question = {"id": "q", "kind": "boolean", "prompt": "Eligible?", "criteria": {"true": {"rules": ["A"]}, "false": "Failure"}}
try:
    local.validate({"schemaVersion": "1", "requestId": "probe", "state": "text", "questions": [question]})
    local_check = "accepted"
except Exception as error:
    local_check = "rejected: " + type(error).__name__
print(json.dumps({
    "networkCalls": 0,
    "coreValidation": checks,
    "macExtraCriteriaValidation": local_check,
    "macOptions": local.options(question),
}, indent=2))
